Multi-Architecture Builds Succeed But Fail to Push

### Description

I'm able to run the build goal successfully, which is able to auth and pull down images.
```
[INFO] --- docker:0.43.4:build (build-docker-image) @ application ---
[INFO] DOCKER> GET unix://127.0.0.1:1/version
[INFO] Building tar: /Users/toconnor/Development/application/target/docker/application-img/tmp/docker-build.tar
[INFO] DOCKER> [application-img:latest] "application": Created docker-build.tar in 570 milliseconds
[INFO] DOCKER> Credentials helper reply for "docker-credential-osxkeychain" is docker-credential-osxkeychain (github.com/docker/docker-credential-helpers) v0.7.0
[INFO] DOCKER> Using Docker CLI 24.0.6
[INFO] DOCKER> docker --config /Users/toconnor/Development/application/target/docker/application-img/docker buildx create --driver docker-container --name multiarch
[INFO] DOCKER> multiarch
[INFO] Expanding: /Users/toconnor/Development/application/target/docker/application-img/tmp/docker-build.tar into /Users/toconnor/Development/application/target/docker/application-img/tmp/docker-build
[INFO] DOCKER> docker --config /Users/toconnor/Development/application/target/docker/application-img/docker buildx build --progress=plain --builder multiarch --platform linux/arm64 --tag application-img:latest --file=/Users/toconnor/Development/application/target/docker/application-img/tmp/docker-build/Dockerfile /Users/toconnor/Development/application/target/docker/application-img/tmp/docker-build --load
[INFO] DOCKER> #0 building with "multiarch" instance using docker-container driver
[INFO] DOCKER>
[INFO] DOCKER> #1 [internal] booting buildkit
[INFO] DOCKER> #1 pulling image moby/buildkit:buildx-stable-1
[INFO] DOCKER> #1 pulling image moby/buildkit:buildx-stable-1 1.4s done
[INFO] DOCKER> #1 creating container buildx_buildkit_multiarch0
[INFO] DOCKER> #1 creating container buildx_buildkit_multiarch0 0.6s done
[INFO] DOCKER> #1 DONE 2.0s
[INFO] DOCKER>
[INFO] DOCKER> #2 [internal] load build definition from Dockerfile
[INFO] DOCKER> #2 transferring dockerfile: 618B done
[INFO] DOCKER> #2 DONE 0.0s
[INFO] DOCKER>
[INFO] DOCKER> #3 [internal] load metadata for docker-repo/asr/java_v17_oracle_build_debug:3
[INFO] DOCKER> #3 ...
[INFO] DOCKER>
[INFO] DOCKER> #4 [auth] asr/java_v17_oracle_build_debug:pull token for docker-repo
[INFO] DOCKER> #4 DONE 0.0s
[INFO] DOCKER>
[INFO] DOCKER> #5 [auth] asr/java_v17_oracle_runtime:pull token for docker-repo
[INFO] DOCKER> #5 DONE 0.0s
[INFO] DOCKER>
[INFO] DOCKER> #6 [internal] load metadata for docker-repo/asr/java_v17_oracle_runtime:2.8.0
[INFO] DOCKER> #6 DONE 1.4s
[INFO] DOCKER>
[INFO] DOCKER> #3 [internal] load metadata for docker-repo/asr/java_v17_oracle_build_debug:3
[INFO] DOCKER> #3 DONE 1.5s
[INFO] DOCKER>
[INFO] DOCKER> #7 [internal] load .dockerignore
[INFO] DOCKER> #7 transferring context: 2B done
[INFO] DOCKER> #7 DONE 0.0s
[INFO] DOCKER>

... 

[INFO] DOCKER>
[INFO] DOCKER> #16 exporting to docker image format
[INFO] DOCKER> #16 exporting layers
[INFO] DOCKER> #16 exporting layers 2.2s done
[INFO] DOCKER> #16 exporting manifest sha256:8e266ee3684ab9ccac16475ed59998e358e2af6f506376bd9618c9542e376d7d 0.0s done
[INFO] DOCKER> #16 exporting config sha256:e8c5bb4c156e81f0d0b110cb66fc1eca0864974eee9a3f8f2e03b63952414a3e
[INFO] DOCKER> #16 exporting config sha256:e8c5bb4c156e81f0d0b110cb66fc1eca0864974eee9a3f8f2e03b63952414a3e 0.0s done
[INFO] DOCKER> #16 sending tarball
[INFO] DOCKER> #16 ...
[INFO] DOCKER>
[INFO] DOCKER> #17 importing to docker
[INFO] DOCKER> #17 DONE 0.7s
[INFO] DOCKER>
[INFO] DOCKER> #16 exporting to docker image format
[INFO] DOCKER> #16 sending tarball 3.6s done
[INFO] DOCKER> #16 DONE 5.8s
```

However, the push goal fails with a 403.
```
[INFO] --- docker:0.43.4:push (push-docker-image) @ application ---
[INFO] DOCKER> GET unix://127.0.0.1:1/version
[INFO] DOCKER> Using Docker CLI 24.0.6
[INFO] Expanding: /Users/toconnor/Development/application/target/docker/application-img/tmp/docker-build.tar into /Users/toconnor/Development/application/target/docker/application-img/tmp/docker-build
[INFO] DOCKER> docker --config /Users/toconnor/Development/application/target/docker/application-img/docker buildx build --progress=plain --builder multiarch --platform linux/amd64,linux/arm64 --tag remote-repo/application-img:latest --file=/Users/toconnor/Development/application/target/docker/application-img/tmp/docker-build/Dockerfile /Users/toconnor/Development/application/target/docker/application-img/tmp/docker-build --push
[INFO] DOCKER> #0 building with "multiarch" instance using docker-container driver
[INFO] DOCKER>
[INFO] DOCKER> #1 [internal] load build definition from Dockerfile
[INFO] DOCKER> #1 transferring dockerfile: 618B done
[INFO] DOCKER> #1 DONE 0.0s
[INFO] DOCKER>
[INFO] DOCKER> #2 [linux/arm64 internal] load metadata for docker-repo/asr/java_v17_oracle_build_debug:3
[INFO] DOCKER> #2 ERROR: unexpected status from HEAD request to https://docker-repo/v2/asr/java_v17_oracle_build_debug/manifests/3: 403 Forbidden
[INFO] DOCKER>
[INFO] DOCKER> #3 [linux/amd64 internal] load metadata for docker-repo/asr/java_v17_oracle_build_debug:3
[INFO] DOCKER> #3 ERROR: unexpected status from HEAD request to https://docker-repo/v2/asr/java_v17_oracle_build_debug/manifests/3: 403 Forbidden
[INFO] DOCKER>
[INFO] DOCKER> #4 [linux/amd64 internal] load metadata for docker-repo/asr/java_v17_oracle_runtime:2.8.0
[INFO] DOCKER> #4 CANCELED
[INFO] DOCKER>
[INFO] DOCKER> #5 [linux/arm64 internal] load metadata for docker-repo/asr/java_v17_oracle_runtime:2.8.0
[INFO] DOCKER> #5 CANCELED
[INFO] DOCKER> ------
[INFO] DOCKER>  > [linux/amd64 internal] load metadata for docker-repo/asr/java_v17_oracle_build_debug:3:
[INFO] DOCKER> ------
[INFO] DOCKER> ------
[INFO] DOCKER>  > [linux/arm64 internal] load metadata for docker-repo/asr/java_v17_oracle_build_debug:3:
[INFO] DOCKER> ------
[INFO] DOCKER> Dockerfile:5
[INFO] DOCKER> --------------------
[INFO] DOCKER>    3 |     ARG jacocoVersion
[INFO] DOCKER>    4 |
[INFO] DOCKER>    5 | >>> FROM ${buildImage} as builder
[INFO] DOCKER>    6 |     ARG jacocoVersion
[INFO] DOCKER> --------------------
[INFO] DOCKER> ERROR: failed to solve: docker-repo/asr/java_v17_oracle_build_debug:3: unexpected status from HEAD request to https://docker-repo/v2/asr/java_v17_oracle_build_debug/manifests/3: 403 Forbidden
[ERROR] DOCKER> Error status (1) when building
```

To make things even more confusing, if I copy the command and execute it locally in the shell the push command works as expected.  Specifically this command which was copied verbatim from the debug output from DMP.
```
docker --config /Users/toconnor/Development/application/target/docker/application-img/docker buildx build --progress=plain --builder multiarch --platform linux/amd64,linux/arm64 --tag remote-repo/application-img:latest --file=/Users/toconnor/Development/application/target/docker/application-img/tmp/docker-build/Dockerfile /Users/toconnor/Development/application/target/docker/application-img/tmp/docker-build --push
```

I've tried every authentication mechanism specified in [the documentation](https://dmp.fabric8.io/#authentication) and they all result in the same behavior.

Both the source registry and destination registry are internally hosted on Artifactory and are v2 Docker registries.  They require auth for all requests.  The fact that I'm able to build successfully proves that I have authentication configured correctly.  On the surface it appears that the HEAD request may not be sending authentication credentials but all the other requests do?

I'm running on an M1 Macbook with the latest Docker Desktop.  As far as I can tell, this seems to be an issue only on Mac or with recent versions of Docker.  This actually works from our Linux AMD64 based CI infrastructure.  However, that infra is using Docker 20.x.x.  I know from other issues reported that there is some variation in behavior between older Docker clients and newer.  

At this point, I'm not sure how to debug this any further.  I'm open to debug builds or even some hints on where to dig into the code.  I spent some time looking at it but there was nothing obvious.

### Info

* docker-maven-plugin version : 0.43.4
* Maven version (`mvn -v`) : 3.9.2
```

```
* Docker version : 
```
Client:
 Cloud integration: v1.0.35+desktop.4
 Version:           24.0.6
 API version:       1.43
 Go version:        go1.20.7
 Git commit:        ed223bc
 Built:             Mon Sep  4 12:28:49 2023
 OS/Arch:           darwin/arm64
 Context:           default

Server: Docker Desktop 4.23.0 (120376)
 Engine:
  Version:          24.0.6
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.7
  Git commit:       1a79695
  Built:            Mon Sep  4 12:31:36 2023
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.6.22
  GitCommit:        8165feabfdfe38c65b599c4993d227328c231fca
 runc:
  Version:          1.1.8
  GitCommit:        v1.1.8-0-g82f18fe
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

* If it's a bug, how to reproduce :

--- a/doc/changelog.md
+++ b/doc/changelog.md
@@ -4,6 +4,7 @@
   - The <noCache> option is now propagated down to the buildx command, if it is set in the <build> section. ([1717](https://github.com/fabric8io/docker-maven-plugin/pull/1717))
   - Fix Buildx build with Dockerfile outside of the Docker build context directory ([1721](https://github.com/fabric8io/docker-maven-plugin/pull/1721))
   - Add support setting driverOpts for buildx ([1704](https://github.com/fabric8io/docker-maven-plugin/pull/1704))
+  - Multi-Architecture push is not sending pull registry auth credentials ([1709](https://github.com/fabric8io/docker-maven-plugin/issues/1709))
 
 * **0.43.4** (2023-08-18):
   - Always pass `--config` option for latest versions of Docker CLI ([1701](https://github.com/fabric8io/docker-maven-plugin/issues/1701))

--- a/src/main/java/io/fabric8/maven/docker/AbstractBuildSupportMojo.java
+++ b/src/main/java/io/fabric8/maven/docker/AbstractBuildSupportMojo.java
@@ -23,24 +23,8 @@ abstract public class AbstractBuildSupportMojo extends AbstractDockerMojo {
     // Parameters required from Maven when building an assembly. They cannot be injected directly
     // into DockerAssemblyCreator.
     // See also here: http://maven.40175.n5.nabble.com/Mojo-Java-1-5-Component-MavenProject-returns-null-vs-JavaDoc-parameter-expression-quot-project-quot-s-td5733805.html
-
-    @Parameter
-    private MavenArchiveConfiguration archive;
-
-    @Component
-    private MavenFileFilter mavenFileFilter;
-
-    @Component
-    private MavenReaderFilter mavenFilterReader;
-
-    @Parameter
-    private Map<String, String> buildArgs;
-
     @Parameter(property = "docker.pull.registry")
     String pullRegistry;
-
-    @Parameter( defaultValue = "${reactorProjects}", required = true, readonly = true )
-    private List<MavenProject> reactorProjects;
 
 
     protected BuildService.BuildContext getBuildContext() throws MojoExecutionException {
@@ -49,11 +33,6 @@ abstract public class AbstractBuildSupportMojo extends AbstractDockerMojo {
                 .mojoParameters(createMojoParameters())
                 .registryConfig(getRegistryConfig(pullRegistry))
                 .build();
-    }
-
-    protected MojoParameters createMojoParameters() {
-        return new MojoParameters(session, project, archive, mavenFileFilter, mavenFilterReader,
-                                  settings, sourceDirectory, outputDirectory, reactorProjects);
     }
 
 }

--- a/src/main/java/io/fabric8/maven/docker/AbstractDockerMojo.java
+++ b/src/main/java/io/fabric8/maven/docker/AbstractDockerMojo.java
@@ -37,9 +37,11 @@ import io.fabric8.maven.docker.util.AuthConfigFactory;
 import io.fabric8.maven.docker.util.EnvUtil;
 import io.fabric8.maven.docker.util.GavLabel;
 import io.fabric8.maven.docker.util.ImageNameFormatter;
+import io.fabric8.maven.docker.util.MojoParameters;
 import io.fabric8.maven.docker.util.NamePatternUtil;
 
 import io.fabric8.maven.docker.util.ProjectPaths;
+import org.apache.maven.archiver.MavenArchiveConfiguration;
 import org.apache.maven.execution.MavenSession;
 import org.apache.maven.plugin.AbstractMojo;
 import org.apache.maven.plugin.MojoExecution;
@@ -49,6 +51,8 @@ import org.apache.maven.plugins.annotations.Component;
 import org.apache.maven.plugins.annotations.Parameter;
 import org.apache.maven.project.MavenProject;
 import org.apache.maven.settings.Settings;
+import org.apache.maven.shared.filtering.MavenFileFilter;
+import org.apache.maven.shared.filtering.MavenReaderFilter;
 import org.apache.maven.shared.utils.logging.MessageUtils;
 import org.codehaus.plexus.PlexusConstants;
 import org.codehaus.plexus.PlexusContainer;
@@ -224,6 +228,20 @@ public abstract class AbstractDockerMojo extends AbstractMojo implements Context
 
     @Parameter(property = "docker.skip.pom", defaultValue = "false")
     protected boolean skipPom;
+
+    @Parameter( defaultValue = "${reactorProjects}", required = true, readonly = true )
+    protected List<MavenProject> reactorProjects;
+    @Parameter
+    protected MavenArchiveConfiguration archive;
+
+    @Component
+    protected MavenFileFilter mavenFileFilter;
+
+    @Component
+    protected MavenReaderFilter mavenFilterReader;
+
+    @Parameter
+    protected Map<String, String> buildArgs;
 
     // Images resolved with external image resolvers and hooks for subclass to
     // mangle the image configurations.
@@ -607,5 +625,10 @@ public abstract class AbstractDockerMojo extends AbstractMojo implements Context
 
     protected ProjectPaths createProjectPaths() {
         return new ProjectPaths(project.getBasedir(), outputDirectory);
+    }
+
+    protected MojoParameters createMojoParameters() {
+        return new MojoParameters(session, project, archive, mavenFileFilter, mavenFilterReader,
+            settings, sourceDirectory, outputDirectory, reactorProjects);
     }
 }

--- a/src/main/java/io/fabric8/maven/docker/BuildMojo.java
+++ b/src/main/java/io/fabric8/maven/docker/BuildMojo.java
@@ -1,5 +1,20 @@
 package io.fabric8.maven.docker;
 
+import io.fabric8.maven.docker.access.DockerAccessException;
+import io.fabric8.maven.docker.config.BuildImageConfiguration;
+import io.fabric8.maven.docker.config.ImageConfiguration;
+import io.fabric8.maven.docker.service.BuildService;
+import io.fabric8.maven.docker.service.ImagePullManager;
+import io.fabric8.maven.docker.service.JibBuildService;
+import io.fabric8.maven.docker.service.ServiceHub;
+import io.fabric8.maven.docker.util.EnvUtil;
+import io.fabric8.maven.docker.util.Logger;
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugins.annotations.LifecyclePhase;
+import org.apache.maven.plugins.annotations.Mojo;
+import org.apache.maven.plugins.annotations.Parameter;
+import org.apache.maven.plugins.annotations.ResolutionScope;
+
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStreamReader;
@@ -7,34 +22,10 @@ import java.io.LineNumberReader;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.net.URL;
-import java.util.Collections;
 import java.util.Date;
 import java.util.Enumeration;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Set;
 
-import io.fabric8.maven.docker.access.AuthConfig;
-import io.fabric8.maven.docker.access.AuthConfigList;
-import io.fabric8.maven.docker.access.DockerAccessException;
-import io.fabric8.maven.docker.config.BuildImageConfiguration;
-import io.fabric8.maven.docker.config.ImageConfiguration;
-import io.fabric8.maven.docker.service.BuildService;
-import io.fabric8.maven.docker.service.ImagePullManager;
-import io.fabric8.maven.docker.service.JibBuildService;
-import io.fabric8.maven.docker.service.RegistryService;
-import io.fabric8.maven.docker.service.ServiceHub;
-import io.fabric8.maven.docker.util.DockerFileUtil;
-import io.fabric8.maven.docker.util.ImageName;
-import io.fabric8.maven.docker.util.Logger;
-import io.fabric8.maven.docker.util.EnvUtil;
-import io.fabric8.maven.docker.util.MojoParameters;
-import org.apache.commons.lang3.StringUtils;
-import org.apache.maven.plugin.MojoExecutionException;
-import org.apache.maven.plugins.annotations.LifecyclePhase;
-import org.apache.maven.plugins.annotations.Mojo;
-import org.apache.maven.plugins.annotations.Parameter;
-import org.apache.maven.plugins.annotations.ResolutionScope;
+import static io.fabric8.maven.docker.service.RegistryService.createCompleteAuthConfigList;
 
 /**
  * Mojo for building a data image
@@ -110,7 +101,7 @@ public class BuildMojo extends AbstractBuildSupportMojo {
         File buildArchiveFile = buildService.buildArchive(imageConfig, buildContext, resolveBuildArchiveParameter());
         if (Boolean.FALSE.equals(shallBuildArchiveOnly())) {
             if (imageConfig.isBuildX()) {
-                hub.getBuildXService().build(createProjectPaths(), imageConfig, null, getAuthConfig(imageConfig), buildArchiveFile);
+                hub.getBuildXService().build(createProjectPaths(), imageConfig, null, createCompleteAuthConfigList(false, imageConfig, getRegistryConfig(pullRegistry), createMojoParameters()), buildArchiveFile);
             } else {
                 buildService.buildImage(imageConfig, pullManager, buildContext, buildArchiveFile);
                 if (!skipTag) {
@@ -120,76 +111,7 @@ public class BuildMojo extends AbstractBuildSupportMojo {
         }
     }
 
-
-    private AuthConfigList getAuthConfig(ImageConfiguration imageConfig) throws MojoExecutionException {
-        // TODO: refactor similar code in RegistryService#pushImages
-        RegistryService.RegistryConfig registryConfig = getRegistryConfig(pullRegistry);
-
-        ImageName imageName = new ImageName(imageConfig.getName());
-        String configuredRegistry = EnvUtil.firstRegistryOf(
-            imageName.getRegistry(),
-            imageConfig.getRegistry(),
-            registryConfig.getRegistry());
-
-        AuthConfig authConfig = registryConfig.createAuthConfig(false, imageName.getUser(), configuredRegistry);
-        AuthConfigList authConfigList = new AuthConfigList();
-        if (authConfig != null) {
-            authConfigList.addAuthConfig(authConfig);
-        }
-
-        BuildImageConfiguration buildConfig = imageConfig.getBuildConfiguration();
-        Set<String> fromRegistries = getRegistriesForPull(buildConfig);
-        for (String fromRegistry : fromRegistries) {
-            if (StringUtils.isNotBlank(configuredRegistry) && configuredRegistry.equalsIgnoreCase(fromRegistry)) {
-                continue;
-            }
-            registryConfig = getRegistryConfig(fromRegistry);
-            AuthConfig additionalAuth = registryConfig.createAuthConfig(false, imageName.getUser(), fromRegistry);
-            if (additionalAuth != null) {
-                authConfigList.addAuthConfig(additionalAuth);
-            }
-        }
-
-        return authConfigList;
-    }
-
-    private Set<String> getRegistriesForPull(BuildImageConfiguration buildConfig) {
-        Set<String> registries = new HashSet<>();
-        List<String> fromImages = extractBaseFromDockerfile(buildConfig);
-        for (String fromImage : fromImages) {
-            ImageName imageName = new ImageName(fromImage);
-
-            if (imageName.hasRegistry()) {
-                registries.add(imageName.getRegistry());
-            }
-        }
-        return registries;
-    }
-
-    private List<String> extractBaseFromDockerfile(BuildImageConfiguration buildConfig) {
-        if (buildConfig.getDockerFile() == null || !buildConfig.getDockerFile().exists()) {
-            if (buildConfig.getFrom() != null && !buildConfig.getFrom().isEmpty()) {
-                return Collections.singletonList(buildConfig.getFrom());
-            }
-            return Collections.emptyList();
-        }
-
-        List<String> fromImage;
-        try {
-            MojoParameters mojoParameters = createMojoParameters();
-            File fullDockerFilePath = buildConfig.getAbsoluteDockerFilePath(mojoParameters);
-            fromImage = DockerFileUtil.extractBaseImages(
-                    fullDockerFilePath,
-                    DockerFileUtil.createInterpolator(mojoParameters, buildConfig.getFilter()),
-                    buildConfig.getArgs());
-        } catch (IOException e) {
-            return Collections.emptyList();
-        }
-        return fromImage;
-    }
-
     // We ignore an already existing date file and always return the current date
-
     @Override
     protected Date getReferenceDate() {
         return new Date();

--- a/src/main/java/io/fabric8/maven/docker/PushMojo.java
+++ b/src/main/java/io/fabric8/maven/docker/PushMojo.java
@@ -51,7 +51,7 @@ public class PushMojo extends AbstractDockerMojo {
     }
 
     private void executeDockerPush(ServiceHub hub) throws MojoExecutionException, DockerAccessException {
-        hub.getRegistryService().pushImages(createProjectPaths(), getResolvedImages(), retries, getRegistryConfig(pushRegistry), skipTag);
+        hub.getRegistryService().pushImages(createProjectPaths(), getResolvedImages(), retries, getRegistryConfig(pushRegistry), skipTag, createMojoParameters());
     }
 
     private void executeJibPush(ServiceHub hub) throws MojoExecutionException {

--- a/src/main/java/io/fabric8/maven/docker/service/RegistryService.java
+++ b/src/main/java/io/fabric8/maven/docker/service/RegistryService.java
@@ -1,9 +1,14 @@
 package io.fabric8.maven.docker.service;
 
+import java.io.File;
+import java.io.IOException;
 import java.io.Serializable;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
 import java.util.Map;
+import java.util.Set;
 
 import io.fabric8.maven.docker.access.AuthConfig;
 import io.fabric8.maven.docker.access.AuthConfigList;
@@ -14,11 +19,14 @@ import io.fabric8.maven.docker.config.BuildImageConfiguration;
 import io.fabric8.maven.docker.config.ImageConfiguration;
 import io.fabric8.maven.docker.config.ImagePullPolicy;
 import io.fabric8.maven.docker.util.AuthConfigFactory;
+import io.fabric8.maven.docker.util.DockerFileUtil;
 import io.fabric8.maven.docker.util.EnvUtil;
 import io.fabric8.maven.docker.util.ImageName;
 import io.fabric8.maven.docker.util.Logger;
 
+import io.fabric8.maven.docker.util.MojoParameters;
 import io.fabric8.maven.docker.util.ProjectPaths;
+import org.apache.commons.lang3.StringUtils;
 import org.apache.maven.plugin.MojoExecutionException;
 import org.apache.maven.settings.Settings;
 
@@ -50,7 +58,7 @@ public class RegistryService {
      * @throws MojoExecutionException
      */
     public void pushImages(ProjectPaths projectPaths, Collection<ImageConfiguration> imageConfigs,
-                            int retries, RegistryConfig registryConfig, boolean skipTag) throws DockerAccessException, MojoExecutionException {
+                            int retries, RegistryConfig registryConfig, boolean skipTag, MojoParameters mojoParameters) throws DockerAccessException, MojoExecutionException {
         for (ImageConfiguration imageConfig : imageConfigs) {
             BuildImageConfiguration buildConfig = imageConfig.getBuildConfiguration();
             if (buildConfig == null || buildConfig.skipPush()) {
@@ -66,12 +74,13 @@ public class RegistryService {
                 imageConfig.getRegistry(),
                 registryConfig.getRegistry());
 
+            AuthConfig authConfigForLegacyPush = createAuthConfig(true, imageName.getUser(), configuredRegistry, registryConfig);
+            AuthConfigList authConfigListForBuildXPush = createCompleteAuthConfigList(true, imageConfig, registryConfig, mojoParameters);
 
-            AuthConfig authConfig = createAuthConfig(true, imageName.getUser(), configuredRegistry, registryConfig);
             if (imageConfig.isBuildX()) {
-                buildXService.push(projectPaths, imageConfig, configuredRegistry, new AuthConfigList(authConfig));
+                buildXService.push(projectPaths, imageConfig, configuredRegistry, authConfigListForBuildXPush);
             } else {
-                dockerPush(retries, skipTag, buildConfig, name, configuredRegistry, authConfig);
+                dockerPush(retries, skipTag, buildConfig, name, configuredRegistry, authConfigForLegacyPush);
             }
         }
     }
@@ -137,8 +146,73 @@ public class RegistryService {
     }
 
 
+    public static AuthConfigList createCompleteAuthConfigList(boolean isPush, ImageConfiguration imageConfig, RegistryConfig registryConfig, MojoParameters mojoParameters) throws MojoExecutionException {
+        ImageName imageName = new ImageName(imageConfig.getName());
+        String configuredRegistry = EnvUtil.firstRegistryOf(
+            imageName.getRegistry(),
+            imageConfig.getRegistry(),
+            registryConfig.getRegistry());
+
+        AuthConfig authConfig = registryConfig.createAuthConfig(isPush, imageName.getUser(), configuredRegistry);
+        AuthConfigList authConfigList = createAuthConfigListForBaseImages(imageConfig.getBuildConfiguration(), mojoParameters, configuredRegistry, registryConfig);
+        if (authConfig != null) {
+            authConfigList.addAuthConfig(authConfig);
+        }
+
+        return authConfigList;
+    }
+
+    public static AuthConfigList createAuthConfigListForBaseImages(BuildImageConfiguration buildConfig, MojoParameters mojoParameters, String configuredRegistry, RegistryConfig registryConfig) throws MojoExecutionException {
+        AuthConfigList authConfigList = new AuthConfigList();
+        Set<String> fromRegistries = getRegistriesForPull(buildConfig, mojoParameters);
+        for (String fromRegistry : fromRegistries) {
+            if (StringUtils.isNotBlank(configuredRegistry) && configuredRegistry.equalsIgnoreCase(fromRegistry)) {
+                continue;
+            }
+            registryConfig.registry = fromRegistry;
+            AuthConfig additionalAuth = registryConfig.createAuthConfig(false, null, fromRegistry);
+            if (additionalAuth != null) {
+                authConfigList.addAuthConfig(additionalAuth);
+            }
+        }
+        return authConfigList;
+    }
+
     // ============================================================================================================
 
+    private static Set<String> getRegistriesForPull(BuildImageConfiguration buildConfig, MojoParameters mojoParameters) {
+        Set<String> registries = new HashSet<>();
+        List<String> fromImages = extractBaseFromDockerfile(buildConfig, mojoParameters);
+        for (String fromImage : fromImages) {
+            ImageName imageName = new ImageName(fromImage);
+
+            if (imageName.hasRegistry()) {
+                registries.add(imageName.getRegistry());
+            }
+        }
+        return registries;
+    }
+
+    private static List<String> extractBaseFromDockerfile(BuildImageConfiguration buildConfig, MojoParameters mojoParameters) {
+        if (buildConfig.getDockerFile() == null || !buildConfig.getDockerFile().exists()) {
+            if (buildConfig.getFrom() != null && !buildConfig.getFrom().isEmpty()) {
+                return Collections.singletonList(buildConfig.getFrom());
+            }
+            return Collections.emptyList();
+        }
+
+        List<String> fromImage;
+        try {
+            File fullDockerFilePath = buildConfig.getAbsoluteDockerFilePath(mojoParameters);
+            fromImage = DockerFileUtil.extractBaseImages(
+                fullDockerFilePath,
+                DockerFileUtil.createInterpolator(mojoParameters, buildConfig.getFilter()),
+                buildConfig.getArgs());
+        } catch (IOException e) {
+            return Collections.emptyList();
+        }
+        return fromImage;
+    }
 
     private boolean imageRequiresPull(boolean hasImage, ImagePullPolicy pullPolicy, String imageName)
         throws MojoExecutionException {

--- a/src/test/java/io/fabric8/maven/docker/PushMojoTest.java
+++ b/src/test/java/io/fabric8/maven/docker/PushMojoTest.java
@@ -2,6 +2,7 @@ package io.fabric8.maven.docker;
 
 import io.fabric8.maven.docker.access.DockerAccessException;
 import io.fabric8.maven.docker.service.RegistryService;
+import io.fabric8.maven.docker.util.MojoParameters;
 import io.fabric8.maven.docker.util.ProjectPaths;
 import org.apache.maven.plugin.MojoExecutionException;
 import org.junit.jupiter.api.Test;
@@ -81,7 +82,7 @@ class PushMojoTest extends MojoTestBase {
 
   private void verifyPush(int wantedNumberOfInvocations) throws DockerAccessException, MojoExecutionException {
     Mockito.verify(registryService, Mockito.times(wantedNumberOfInvocations))
-        .pushImages(Mockito.any(ProjectPaths.class), Mockito.anyCollection(), Mockito.anyInt(), Mockito.any(RegistryService.RegistryConfig.class), Mockito.anyBoolean());
+        .pushImages(Mockito.any(ProjectPaths.class), Mockito.anyCollection(), Mockito.anyInt(), Mockito.any(RegistryService.RegistryConfig.class), Mockito.anyBoolean(), Mockito.any(MojoParameters.class));
   }
 
   private void whenMojoExecutes() throws IOException, MojoExecutionException {

--- a/src/test/java/io/fabric8/maven/docker/service/RegistryServicePushImagesBuildXTest.java
+++ b/src/test/java/io/fabric8/maven/docker/service/RegistryServicePushImagesBuildXTest.java
@@ -1,0 +1,143 @@
+package io.fabric8.maven.docker.service;
+
+import io.fabric8.maven.docker.access.AuthConfig;
+import io.fabric8.maven.docker.access.AuthConfigList;
+import io.fabric8.maven.docker.access.DockerAccess;
+import io.fabric8.maven.docker.access.DockerAccessException;
+import io.fabric8.maven.docker.config.BuildImageConfiguration;
+import io.fabric8.maven.docker.config.BuildXConfiguration;
+import io.fabric8.maven.docker.config.ImageConfiguration;
+import io.fabric8.maven.docker.util.AuthConfigFactory;
+import io.fabric8.maven.docker.util.Logger;
+import io.fabric8.maven.docker.util.ProjectPaths;
+import org.apache.maven.plugin.MojoExecutionException;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+import org.mockito.ArgumentCaptor;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.StandardOpenOption;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyBoolean;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+class RegistryServicePushImagesBuildXTest {
+  private RegistryService registryService;
+  private BuildXService buildXService;
+  private List<ImageConfiguration> imageConfigurationList;
+  private ProjectPaths projectPaths;
+  private RegistryService.RegistryConfig registryConfig;
+  private AuthConfigFactory authConfigFactory;
+
+  @TempDir
+  private File temporaryFolder;
+
+  @BeforeEach
+  void setUp() {
+    buildXService = mock(BuildXService.class);
+    Logger logger = mock(Logger.class);
+    authConfigFactory = mock(AuthConfigFactory.class);
+    DockerAccess dockerAccess = mock(DockerAccess.class);
+    QueryService queryService = new QueryService(dockerAccess);
+    imageConfigurationList = Collections.singletonList(createNewImageConfiguration("user1/sample-image:latest", "foo/base:latest", null));
+    registryConfig = new RegistryService.RegistryConfig.Builder()
+        .registry("registry1.org")
+        .authConfigFactory(authConfigFactory)
+        .build();
+    projectPaths = new ProjectPaths(temporaryFolder, "target/docker");
+    registryService = new RegistryService(dockerAccess, queryService, buildXService, logger);
+  }
+
+  @Test
+  void whenNoRegistryConfigured_thenAuthConfigEmpty() throws MojoExecutionException, DockerAccessException {
+    // When
+    registryService.pushImages(projectPaths, imageConfigurationList, 0, registryConfig, false, null);
+
+    // Then
+    verifyBuildXServiceInvokedWithAuthConfigListSize(0);
+  }
+
+  @Test
+  void whenOnlyPushRegistryConfigured_thenAuthConfigHasSingleEntry() throws MojoExecutionException, DockerAccessException {
+    // Given
+    givenAuthConfigExistsForRegistry("registry1.org", "user1", "password1");
+
+    // When
+    registryService.pushImages(projectPaths, imageConfigurationList, 0, registryConfig, false, null);
+
+    // Then
+    verifyBuildXServiceInvokedWithAuthConfigListSize(1);
+  }
+
+  @Test
+  void whenFromRegistryAndPushRegistryProvided_thenAuthConfigListContainsEntriesForBothPullAndPush() throws MojoExecutionException, DockerAccessException {
+    // Given
+    imageConfigurationList = Collections.singletonList(createNewImageConfiguration("user1/base-image-different-registry:latest", "registry2.org/user2/base:latest", null));
+    givenAuthConfigExistsForRegistry("registry1.org", "user1", "password1");
+    givenAuthConfigExistsForRegistry("registry2.org", "user2", "password2");
+
+    // When
+    registryService.pushImages(projectPaths, imageConfigurationList, 0, registryConfig, false, null);
+
+    // Then
+    verifyBuildXServiceInvokedWithAuthConfigListSize(2);
+  }
+
+  @Test
+  void whenDockerfileContainsFromReferencingMultipleRegistries_thenAuthConfigListContainsMultipleEntries() throws MojoExecutionException, IOException {
+    // Given
+    File dockerFile = new File(temporaryFolder, "Dockerfile");
+    Files.write(dockerFile.toPath(), String.format("FROM registry2.org/user2/test-base:latest%n").getBytes());
+    Files.write(dockerFile.toPath(), "FROM registry3.org/user3/scratch:1.0 AS build2".getBytes(), StandardOpenOption.APPEND);
+    imageConfigurationList = Collections.singletonList(createNewImageConfiguration("user1/base-image-different-registry:latest", null, dockerFile));
+    givenAuthConfigExistsForRegistry("registry1.org", "user1", "password1");
+    givenAuthConfigExistsForRegistry("registry2.org", "user2", "password2");
+    givenAuthConfigExistsForRegistry("registry3.org", "user3", "password2");
+
+    // When
+    registryService.pushImages(projectPaths, imageConfigurationList, 0, registryConfig, false, null);
+
+    // Then
+    verifyBuildXServiceInvokedWithAuthConfigListSize(3);
+  }
+
+  private void verifyBuildXServiceInvokedWithAuthConfigListSize(int expectedSize) throws MojoExecutionException {
+    ArgumentCaptor<AuthConfigList> authConfigListArgumentCaptor = ArgumentCaptor.forClass(AuthConfigList.class);
+    verify(buildXService).push(any(), any(), anyString(), authConfigListArgumentCaptor.capture());
+    assertEquals(expectedSize, authConfigListArgumentCaptor.getValue().size());
+  }
+
+  private void givenAuthConfigExistsForRegistry(String registry, String username, String password) throws MojoExecutionException {
+    AuthConfig auth = new AuthConfig(username, password, null, null);
+    auth.setRegistry(registry);
+
+    when(authConfigFactory.createAuthConfig(anyBoolean(), anyBoolean(), any(), any(), any(), eq(registry)))
+        .thenReturn(auth);
+  }
+
+  private ImageConfiguration createNewImageConfiguration(String name, String from, File dockerFile) {
+    BuildImageConfiguration buildImageConfiguration = mock(BuildImageConfiguration.class);
+    when(buildImageConfiguration.getFrom()).thenReturn(from);
+    when(buildImageConfiguration.getBuildX()).thenReturn(new BuildXConfiguration.Builder()
+        .platforms(Arrays.asList("linux/amd64", "linux/arm64"))
+        .build());
+    when(buildImageConfiguration.isBuildX()).thenReturn(true);
+    when(buildImageConfiguration.getDockerFile()).thenReturn(dockerFile);
+    when(buildImageConfiguration.getAbsoluteDockerFilePath(any())).thenReturn(dockerFile);
+    return new ImageConfiguration.Builder()
+        .name(name)
+        .buildConfig(buildImageConfiguration).build();
+  }
+}

--- a/src/test/java/io/fabric8/maven/docker/service/RegistryServiceTest.java
+++ b/src/test/java/io/fabric8/maven/docker/service/RegistryServiceTest.java
@@ -519,7 +519,7 @@ class RegistryServiceTest {
                     .authConfig(authConfig)
                     .registry(registry)
                     .build();
-            registryService.pushImages(projectPaths, Collections.singleton(imageConfiguration), 1, registryConfig, false);
+            registryService.pushImages(projectPaths, Collections.singleton(imageConfiguration), 1, registryConfig, false, null);
         } catch (Exception e) {
             this.actualException = e;
         }

